cdap-ranger/cdap-ranger-binding/src/main/java/co/cask/cdap/security/authorization/ranger/binding/RangerAuthorizer.java

```diff
@@ -0,0 +1,301 @@
+import co.cask.cdap.security.spi.authorization.Authorizer;
+import co.cask.cdap.security.spi.authorization.UnauthorizedException;
+import com.google.common.base.Preconditions;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.ranger.audit.provider.MiscUtil;
+import org.apache.ranger.plugin.audit.RangerDefaultAuditHandler;
+import org.apache.ranger.plugin.policyengine.RangerAccessRequestImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessResourceImpl;
+import org.apache.ranger.plugin.policyengine.RangerAccessResult;
+import org.apache.ranger.plugin.service.RangerBasePlugin;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.net.InetAddress;
+import java.util.Date;
+import java.util.Properties;
+import java.util.Set;
+
+/**
+ * This class implements {@link Authorizer} from CDAP and is responsible for interacting with Ranger to enforce
+ * authorization.
+ */
+public class RangerAuthorizer extends AbstractAuthorizer {
+  private static final Logger LOG = LoggerFactory.getLogger(RangerAuthorizer.class);
+
+  // just string keys used to store entity in ranger. We don't want them to be derived from entity type or name since
+  // any changes to them on cdap side will make privileges incompatible.
+  private static final String KEY_INSTANCE = "instance";
+  private static final String KEY_NAMESPACE = "namespace";
+  private static final String KEY_ARTIFACT = "artifact";
+  private static final String KEY_APPLICATION = "application";
+  private static final String KEY_DATASET = "dataset";
+  private static final String KEY_STREAM = "stream";
+  private static final String KEY_PROGRAM = "program";
+  private static final String KEY_DATASET_MODULE = "dataset_module";
+  private static final String KEY_DATASET_TYPE = "dataset_type";
+  private static final String KEY_SECUREKEY = "securekey";
+
+  // using # as we don't allow it in entity names
+  private static final String RESOURCE_SEPARATOR = "#";
+
+  private static volatile RangerBasePlugin rangerPlugin = null;
+  private AuthorizationContext context;
+  // cdap instance name
+  private String instanceName;
+
+  @Override
+  public synchronized void initialize(AuthorizationContext context) throws Exception {
+    this.context = context;
+    Properties properties = context.getExtensionProperties();
+    instanceName = properties.containsKey(Constants.INSTANCE_NAME) ?
+      properties.getProperty(Constants.INSTANCE_NAME) : "cdap";
+    if (rangerPlugin == null) {
+      try {
+        UserGroupInformation ugi = UserGroupInformation.getLoginUser();
+        Preconditions.checkNotNull(ugi, "Kerberos login information is not available. UserGroupInformation is null");
+        // set the login user as the user as whom cdap is running as this is needed for kerberos authentication
+        MiscUtil.setUGILoginUser(ugi, null);
+        LOG.debug("Initializing Ranger CDAP Plugin with UGI {}", ugi);
+      } catch (Throwable t) {
```

Comment:
Whats the rational behind it ? Are we deciding that extension initialization failure should cause cdap master start up  failure ?